crossplane/crossplane
# Migrate from Policy MergeOptions to ToFieldPath in `beta convert pipeline-composition`

### Description of your changes



Fixes #5421 

ℹ️  **Key implementation note**: The types in this PR follow a consistent pattern of a struct that embeds the underlying `Patch` type (that contains the old policy format) while also inserting a new `Policy` field at the same level of this embedded `Patch` struct. When the converted data is serialized to the final converted output manifest, this has the effect of essentially overriding the deprecated policy format with our new policy format.

### Useful resources for reviewers:

* Definition of what function-p-and-t supports and how it differs: https://github.com/crossplane-contrib/function-patch-and-transform/?tab=readme-ov-file#differences-from-the-native-implementation
* @negz suggested design for this implementation: https://github.com/crossplane-contrib/function-patch-and-transform/issues/101#issuecomment-2024155035
* PR that updated function-p-and-t: https://github.com/crossplane-contrib/function-patch-and-transform/pull/102/files

Here is a logic table that captures the conversion rules we need to codify in this function. I've kept the case ordering @negz used in his linked comments above for consistency:

| # | `MergeOptions` | `appendSlice`   | `keepMapValues` | `policy.toFieldPath` |
| -----------  | -----------  | ------------- | -----------  | ------------- |
| 1 | `nil`  | N/A | N/A | `nil` which defaults to `Replace` |
| 2 | `non-nil` | `nil` or `false` | `true` | `MergeObjects` | 
| 3 | `non-nil` | `true` | `nil` or `false` | `ForceMergeObjectsAppendArrays` |
| 4 | `non-nil` | `nil` or `false` |  `nil` or `false` |  `ForceMergeObjects`   |
| 5 | `non-nil` | `true` | `true` | `MergeObjectsAppendArrays` |

### Manual testing

A gist containing test manifests and expected output can be found at https://gist.github.com/jbw976/1c06224c0edc4cf9ca9a3765539a1e7f

In its source composition file, it contains patches for all the possible `mergeOptions` combinations in all the possible locations (environment, patchsets, resources). The converted composition has been verified to contain properly converted policies in their correct locations. As a further validation step, the converted composition has been used with the `render` command to sanity test its structure and content.

I have: 

  - https://github.com/crossplane-contrib/function-patch-and-transform/pull/112 

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet
